use overlayfs (resolves #6 and probably #9)

Nemo64 opened the pull request:
Through this change I made the entire /var folder writable without needing to copy anything.

It'll even work with all bundles that somehow write into the var folder since there is no detectable difference to a real writable folder, even when accessed though command and overlayfs is heavily tested since it is one of the core components of docker.

Nemo64 commented:
This solution works very well, my first request is at `1192.97 ms` down from `1254.10 ms` with obviously no time to copy cache files.

But it is kind of ugly to mount a filesystem during a request. Also this isn't really symfony specific.
@mnapoli I think this could be a feature of the bref runtime. I'm thinking of checking an environment variable like `BREF_OVERLAYFS=/var` which would mount the `/var` dir into `/tmp`. This could be useful for many projects that use some form of cache that isn't designed to be read-only.

Nyholm commented:
Oh, hm. 

I agree with you. Doing this mount here rings all kinds of warnings.  =)

Looking at your first request times.. Isnt quite slow? How long does it take to mount the file system?

Nemo64 commented:
Less than the 82 ms it took before to copy the cache folder.
But to be sure I retested by removing my cache and warming it freshly and dumping the autoload with authorative classmap and I added microtime around the mount command.

Result:
First request took `1191.21 ms` (basically the same) and the mount took `4.76 ms`.
The next requests took `56.46 ms` to give you an idea on the complexity of the page i'm loading.
Here is the log:
```
START RequestId: b2c14d7d-5ff3-424c-90eb-46dd565cbca2 Version: $LATEST
[11-Apr-2020 19:23:05] NOTICE: fpm is running, pid 9
[11-Apr-2020 19:23:05] NOTICE: ready to handle connections
sh: mount: command not found
[2020-04-11T21:23:05+02:00] Mount took 4.76 ms.
[2020-04-11T21:23:05.973048+02:00] request.INFO: Matched route "login". {"route":"login","route_parameters":{"_route":"login","_controller":"App\\Controller\\Facade\\LoginController::login"},"request_uri":"https://dev.symfony-bootstrap.hauptsache.net/login","method":"GET"} []
[2020-04-11T21:23:06.254861+02:00] security.INFO: Populated the TokenStorage with an anonymous Token. [] []
END RequestId: b2c14d7d-5ff3-424c-90eb-46dd565cbca2
REPORT RequestId: b2c14d7d-5ff3-424c-90eb-46dd565cbca2	Duration: 1191.21 ms	Billed Duration: 1600 ms	Memory Size: 1024 MB	Max Memory Used: 126 MB	Init Duration: 353.37 ms	

START RequestId: 2d0f1c1d-c8c3-45a1-8357-bd77f7181bff Version: $LATEST
[2020-04-11T21:23:07.892792+02:00] request.INFO: Matched route "login". {"route":"login","route_parameters":{"_route":"login","_controller":"App\\Controller\\Facade\\LoginController::login"},"request_uri":"https://dev.symfony-bootstrap.hauptsache.net/login","method":"GET"} []
[2020-04-11T21:23:07.931346+02:00] security.INFO: Populated the TokenStorage with an anonymous Token. [] []
END RequestId: 2d0f1c1d-c8c3-45a1-8357-bd77f7181bff
REPORT RequestId: 2d0f1c1d-c8c3-45a1-8357-bd77f7181bff	Duration: 56.46 ms	Billed Duration: 100 ms	Memory Size: 1024 MB	Max Memory Used: 126 MB	
```

Then I switched back to the current master with those results:
First request took `1270.10 ms` with cache copy time of `82.87 ms`.
The next requests took `59.13 ms` which is similar~isch to the overlayfs variant if not marginally slower.
```
START RequestId: 5f9cbca2-6dc4-4002-aa5b-9901fa07c8f4 Version: $LATEST
[11-Apr-2020 19:26:19] NOTICE: fpm is running, pid 8
[11-Apr-2020 19:26:19] NOTICE: ready to handle connections
[2020-04-11T21:26:19+02:00] Symfony cache directory prepared in 82.87 ms.
[2020-04-11T21:26:20.398498+02:00] request.INFO: Matched route "login". {"route":"login","route_parameters":{"_route":"login","_controller":"App\\Controller\\Facade\\LoginController::login"},"request_uri":"https://dev.symfony-bootstrap.hauptsache.net/login","method":"GET"} []
[2020-04-11T21:26:20.691723+02:00] security.INFO: Populated the TokenStorage with an anonymous Token. [] []
END RequestId: 5f9cbca2-6dc4-4002-aa5b-9901fa07c8f4
REPORT RequestId: 5f9cbca2-6dc4-4002-aa5b-9901fa07c8f4	Duration: 1270.10 ms	Billed Duration: 1700 ms	Memory Size: 1024 MB	Max Memory Used: 131 MB	Init Duration: 356.43 ms	

START RequestId: d6b16318-24f6-4e8b-a081-6db56f5bf6d5 Version: $LATEST
[2020-04-11T21:26:22.063656+02:00] request.INFO: Matched route "login". {"route":"login","route_parameters":{"_route":"login","_controller":"App\\Controller\\Facade\\LoginController::login"},"request_uri":"https://dev.symfony-bootstrap.hauptsache.net/login","method":"GET"} []
[2020-04-11T21:26:22.102903+02:00] security.INFO: Populated the TokenStorage with an anonymous Token. [] []
END RequestId: d6b16318-24f6-4e8b-a081-6db56f5bf6d5
REPORT RequestId: d6b16318-24f6-4e8b-a081-6db56f5bf6d5	Duration: 59.13 ms	Billed Duration: 100 ms	Memory Size: 1024 MB	Max Memory Used: 131 MB	
```

That combined with the reduced complexity and the fact that overlayfs does not need implementation aware workarounds makes this the ideal solution imo. Even shell scripts will work just as expected.
I tried to find some documentation around mounting overlayfs within lambda and haven't found anything. That could be the biggest downside, that this isn't an official way.

Nyholm commented:
Sorry for asking stupid questions. But why do the first requests take so long? If the mount is just 4ms shouldn’t the first request be just around 60ms?

Also, why is this line in the logs?

> sh: mount: command not found

Nemo64 commented:
Good question, i'll investigate. Maybe I have some deeper problem in my setup.

Nemo64 commented:
Ok, it seems my application has a cache warmed so far that it mostly runs without write access to `/var/`. If i continue to click around i get write warnings that cache can't be written but those aren't fatal. 
I also have apcu enabled so symfony makes the app.cache a `ChainCache` with `ApcuCache` in front of `FileCache`.

I totally overlooked that the command fails. My application previously just failed when it couldn't write so I saw it working as "the mount must have worked"
